Unit test maintenance re legacy disparity

Prior to embarking on our Rockstor 4 release candidates the existing unit tests on our prior CentOS base were modified to have zero failures/errors in the following pull request:
"remark out all currently failing unit tests. Fixes #2137" #2153
In the mean time we have added additional unit tests and still have some existing failures in prior tests on the non legacy 'Built on openSUSE' endeavour. This issue is to address this disparity by returning our unit test status to that of our prior OS base with the intention of establishing this state prior to our final Stable 4 release.

We have 2 failures in the user section which seem to relate only to differing default users:
./bin/test -v 2 -p test_user*
> Ran 8 tests in 1.578s
> FAILED (failures=2)

details:

> ======================================================================
> FAIL: test_delete_requests (storageadmin.tests.test_user.UserTests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/opt/rockstor-dev/src/rockstor/storageadmin/tests/test_user.py", line 399, in test_delete_requests
>     status.HTTP_200_OK, msg=response.data)
> AssertionError: ['User (games) does not exist.', 'None\n']
> 
> ======================================================================
> FAIL: test_post_requests (storageadmin.tests.test_user.UserTests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/opt/rockstor-dev/src/rockstor/storageadmin/tests/test_user.py", line 175, in test_post_requests
>     msg=response.data)
> AssertionError: {'username': u'newUser', 'public_key': None, 'shell': u'/bin/bash', 'group': 2, 'pincard_allowed': 'no', 'admin': True, 'managed_user': True, 'homedir': u'/home/newUser', 'email': None, 'groupname': u'admin', 'gid': 5, 'user': 8, 'uid': 3, 'smb_shares': [], u'id': 1, 'has_pincard': False}

2 failed in commands likely due to missing files associated with yum auto update:
./bin/test -v 2 -p test_commands*

> Ran 17 tests in 2.976s
> FAILED (failures=2)

details

> ======================================================================
> FAIL: test_disable_auto_update_command (storageadmin.tests.test_commands.CommandTests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/opt/rockstor-dev/src/rockstor/storageadmin/tests/test_commands.py", line 152, in test_disable_auto_update_command
>     status.HTTP_200_OK, msg=response.data)
> AssertionError: ["Failed to disable auto update due to this exception:  ([Errno 2] No such file or directory: '/etc/yum/yum-cron.conf').", 'Traceback (most recent call last):\n  File "/opt/rockstor-dev/src/rockstor/storageadmin/views/command.py", line 392, in post\n    auto_update(enable=False)\n  File "/opt/rockstor-dev/src/rockstor/system/pkg_mgmt.py", line 62, in auto_update\n    with open(YCFILE) as ifo, open(npath, "w") as tfo:\nIOError: [Errno 2] No such file or directory: \'/etc/yum/yum-cron.conf\'\n']
> 
> ======================================================================
> FAIL: test_enable_auto_update_command (storageadmin.tests.test_commands.CommandTests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/opt/rockstor-dev/src/rockstor/storageadmin/tests/test_commands.py", line 144, in test_enable_auto_update_command
>     status.HTTP_200_OK, msg=response.data)
> AssertionError: ["Failed to enable auto update due to this exception: ([Errno 2] No such file or directory: '/etc/yum/yum-cron.conf').", 'Traceback (most recent call last):\n  File "/opt/rockstor-dev/src/rockstor/storageadmin/views/command.py", line 382, in post\n    auto_update(enable=True)\n  File "/opt/rockstor-dev/src/rockstor/system/pkg_mgmt.py", line 62, in auto_update\n    with open(YCFILE) as ifo, open(npath, "w") as tfo:\nIOError: [Errno 2] No such file or directory: \'/etc/yum/yum-cron.conf\'\n']

1 failing in user group:
./bin/test -v 2 -p test_group*

> Ran 3 tests in 0.687s
> FAILED (failures=1)

details

> ======================================================================
> FAIL: test_post_requests (storageadmin.tests.test_group.GroupTests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/opt/rockstor-dev/src/rockstor/storageadmin/tests/test_group.py", line 99, in test_post_requests
>     msg=response.data)
> AssertionError: {'admin': True, 'groupname': u'ngroup2', 'gid': 1, u'id': 1}
> 
> ----------------------------------------------------------------------

Don't see how this passed previously but haven't really looked yet.

And the last failure is in a new networking test that is suspected as having incomplete mocking re docker, or a reliance on the test system running docker or the like.
./bin/test -v 2 -p test_system_network*

> Ran 6 tests in 0.045s
> FAILED (failures=1)

details:

> ======================================================================
> FAIL: test_get_con_config (system.tests.test_system_network.SystemNetworkTests)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "/opt/rockstor-dev/src/rockstor/system/tests/test_system_network.py", line 690, in test_get_con_config
>     "expected = {}.\n ".format(returned, expected),
> AssertionError: Un-expected get_con_config() result:
>  returned = {'7ffa9c37-4559-4608-80de-ea77a27876cd': {'bridge': {'aux_address': None, 'subnet': None, 'internal': False, 'host_binding': None, 'docker_name': None, 'icc': False, 'ip_masquerade': False, 'dgateway': None, 'ip_range': None}, 'ctype': 'bridge', 'ipv6_addresses': None, 'ipv4_method': 'manual', 'ipv6_method': None, 'ipv6_dns': None, 'name': 'docker0', 'ipv4_addresses': '172.17.0.1/16', 'ipv6_gw': None, 'ipv4_dns': None, 'state': 'activated', 'ipv6_dns_search': None, 'ipv4_gw': None, 'ipv4_dns_search': None}}.
>  expected = {'7ffa9c37-4559-4608-80de-ea77a27876cd': {'bridge': {'aux_address': None, 'subnet': u'172.17.0.0/16', 'internal': False, 'host_binding': u'0.0.0.0', 'docker_name': 'docker0', 'icc': u'true', 'ip_masquerade': False, 'dgateway': u'172.17.0.1', 'ip_range': None}, 'ctype': 'bridge', 'ipv6_addresses': None, 'ipv4_method': 'manual', 'ipv6_method': None, 'ipv6_dns': None, 'name': 'docker0', 'ipv4_addresses': '172.17.0.1/16', 'ipv6_gw': None, 'ipv4_dns': None, 'state': 'activated', 'ipv6_dns_search': None, 'ipv4_gw': None, 'ipv4_dns_search': None}}.
>  
> 
> ----------------------------------------------------------------------

Only the recent added network failure one to go now:
`./bin/test -v 2 -p test_*
`

> Ran 212 tests in 37.620s
> FAILED (failures=1)

Details:

```
FAIL: test_get_con_config (system.tests.test_system_network.SystemNetworkTests)

Traceback (most recent call last):
  File "/opt/rockstor-dev/src/rockstor/system/tests/test_system_network.py", line 690, in test_get_con_config
    "expected = {}.\n ".format(returned, expected),
AssertionError: Un-expected get_con_config() result:
 returned = {'7ffa9c37-4559-4608-80de-ea77a27876cd': {'bridge': {'aux_address': None, 'subnet': None, 'internal': False, 'host_binding': None, 'docker_name': None, 'icc': False, 'ip_masquerade': False, 'dgateway': None, 'ip_range': None}, 'ctype': 'bridge', 'ipv6_addresses': None, 'ipv4_method': 'manual', 'ipv6_method': None, 'ipv6_dns': None, 'name': 'docker0', 'ipv4_addresses': '172.17.0.1/16', 'ipv6_gw': None, 'ipv4_dns': None, 'state': 'activated', 'ipv6_dns_search': None, 'ipv4_gw': None, 'ipv4_dns_search': None}}.
 expected = {'7ffa9c37-4559-4608-80de-ea77a27876cd': {'bridge': {'aux_address': None, 'subnet': u'172.17.0.0/16', 'internal': False, 'host_binding': u'0.0.0.0', 'docker_name': 'docker0', 'icc': u'true', 'ip_masquerade': False, 'dgateway': u'172.17.0.1', 'ip_range': None}, 'ctype': 'bridge', 'ipv6_addresses': None, 'ipv4_method': 'manual', 'ipv6_method': None, 'ipv6_dns': None, 'name': 'docker0', 'ipv4_addresses': '172.17.0.1/16', 'ipv6_gw': None, 'ipv4_dns': None, 'state': 'activated', 'ipv6_dns_search': None, 'ipv4_gw': None, 'ipv4_dns_search': None}}.
```

--- a/src/rockstor/storageadmin/tests/test_commands.py
+++ b/src/rockstor/storageadmin/tests/test_commands.py
@@ -56,6 +56,12 @@ class CommandTests(APITestMixin, APITestCase):
         cls.mock_update_check = cls.patch_update_check.start()
         cls.mock_update_check.return_value = 1, 1, 1
 
+        # Mock src/rockstor/system/pkg_mgmt.py auto_update() in
+        # src/rockstor/storageadmin/views/command.py
+        cls.patch_auto_update = patch('storageadmin.views.command.auto_update')
+        cls.mock_auto_update = cls.patch_auto_update.start()
+        cls.mock_auto_update.return_value = True
+
         cls.patch_system_shutdown = patch(
             'storageadmin.views.command.system_shutdown')
         cls.mock_system_shutdown = cls.patch_system_shutdown.start()

--- a/src/rockstor/storageadmin/tests/test_group.py
+++ b/src/rockstor/storageadmin/tests/test_group.py
@@ -92,12 +92,12 @@ class GroupTests(APITestMixin, APITestCase):
 
         # invalid gid
         data = {'groupname': 'ngroup2',
-                'gid': 1001, }
+                'gid': 100, }
         response = self.client.post(self.BASE_URL, data=data)
         self.assertEqual(response.status_code,
                          status.HTTP_400_BAD_REQUEST,
                          msg=response.data)
-        err_msg = 'GID (1001) already exists. Choose a different one.'
+        err_msg = 'GID (100) already exists. Choose a different one.'
         self.assertEqual(response.data[0], err_msg)
 
         # happy path

--- a/src/rockstor/storageadmin/tests/test_samba.py
+++ b/src/rockstor/storageadmin/tests/test_samba.py
@@ -347,12 +347,24 @@ class SambaTests(APITestMixin, APITestCase, SambaListView):
         e_msg = "Must provide share names."
         self.assertEqual(response.data[0], e_msg)
 
+    @mock.patch("storageadmin.models.user.grp.getgrgid")
     @mock.patch("storageadmin.views.samba.ShareMixin._validate_share")
     @mock.patch("storageadmin.views.samba.User")
-    def test_post_requests_2(self, mock_user, mock_validate_share):
+    def test_post_requests_2(self, mock_user, mock_validate_share, mock_getgrgid):
         """
          . Create a samba export for the share that has already been exported
         """
+        # Nullify exception of getgrgid via:
+        mock_getgrgid.side_effects = None
+
+        # Python2
+        # import grp
+        # >>> groupname = grp.getgrgid("470")
+        # >>> print(groupname)
+        # grp.struct_group(gr_name='ntp', gr_passwd='x', gr_gid=470, gr_mem=[])
+        # return "testgroup" for all gr_name calls:
+        mock_getgrgid.gr_name.return_value = "testgroup"
+
         mock_validate_share.return_value = self.temp_share_smb
 
         # create samba with invalid browsable, guest_ok, read_only choices
@@ -413,6 +425,9 @@ class SambaTests(APITestMixin, APITestCase, SambaListView):
         }
         mock_validate_share.return_value = self.temp_share2
         mock_user.objects.get.side_effects = None
+        # The following relies on our mocked return value in grp.getgrgid.gr_name.
+        # Otherwise we error/exception out with grp.getgrgid(1) in models/user.py
+        # 'getgrgid(): gid not found: 1'
         temp_user = User.objects.create(
             username="admin", uid=1, gid=1, admin=False, user=self.user
         )

--- a/src/rockstor/storageadmin/tests/test_user.py
+++ b/src/rockstor/storageadmin/tests/test_user.py
@@ -165,15 +165,15 @@ class UserTests(APITestMixin, APITestCase):
         #          "username.")
         # self.assertEqual(response.data[0], e_msg)
 
-        # create a user with existing uid ('nobody' has a uid 99)
+        # create a user with existing uid ('nobody' has a uid 65534)
         # TODO: We are not limiting user id to >= 1000.
         data = {'username': 'newUser', 'password': 'pwuser2',
-                'group': 'admin', 'uid': '99', 'pubic_key': 'xxx'}
+                'group': 'admin', 'uid': '65534', 'pubic_key': 'xxx'}
         response = self.client.post(self.BASE_URL, data=data)
         self.assertEqual(response.status_code,
                          status.HTTP_500_INTERNAL_SERVER_ERROR,
                          msg=response.data)
-        e_msg = "UID (99) already exists. Please choose a different one."
+        e_msg = "UID (65534) already exists. Please choose a different one."
         self.assertEqual(response.data[0], e_msg)
 
         # create a user that is already a system user(eg: root)
@@ -390,7 +390,10 @@ class UserTests(APITestMixin, APITestCase):
                          status.HTTP_500_INTERNAL_SERVER_ERROR,
                          msg=response.data)
 
-        username = 'games'
+        # Hack fix as lp exists as real system user but is currently, incorrectly, not
+        # included in the exclude_list within src/rockstor/storageadmin/views/user.py
+        # Proper approach would be to setup a User.objects mock
+        username = 'lp'
         response = self.client.delete('{}/{}'.format(self.BASE_URL, username))
         self.assertEqual(response.status_code,
                          status.HTTP_200_OK, msg=response.data)

--- a/src/rockstor/system/tests/test_pkg_mgmt.py
+++ b/src/rockstor/system/tests/test_pkg_mgmt.py
@@ -627,3 +627,11 @@ class SystemPackageTests(unittest.TestCase):
                 "returned = ({}).\n "
                 "expected = ({}).".format(returned, expected),
             )
+
+    # TODO: Add test for zypper auto update functionality post transition from currently
+    #  legacy yum based system which was previously lightly tested to exist only via
+    #  API level unit test src/rockstor/storageadmin/tests/test_commands.py
+    #  Suggested name:
+    #  test_auto_update(self)
+    #  Related additional low level test:
+    #  test_auto_update_status(self)

--- a/src/rockstor/system/tests/test_system_network.py
+++ b/src/rockstor/system/tests/test_system_network.py
@@ -163,8 +163,13 @@ class SystemNetworkTests(unittest.TestCase):
         which should return a dict with detailed config for each network connection detected.
         """
         # Mock and patch docker-specific calls
+        self.patch_docker_status = patch("system.network.docker_status")
+        self.mock_docker_status = self.patch_docker_status.start()
+        self.mock_docker_status.return_value = True
+
         self.patch_dnets = patch("system.network.dnets")
         self.mock_dnets = self.patch_dnets.start()
+
         self.patch_dnet_inspect = patch("system.network.dnet_inspect")
         self.mock_dnet_inspect = self.patch_dnet_inspect.start()
 

